File: examples/crypto/secp256k1/transaction-signing.zig
/// Transaction signing and verification
///
/// Demonstrates:
/// - Legacy transaction signing (pre-EIP-1559)
/// - EIP-155 replay protection
/// - Transaction hash computation
/// - Sender recovery from transaction signature
/// - Transaction verification

const std = @import("std");
const crypto = @import("crypto");
const primitives = @import("primitives");

fn deriveAddress(allocator: std.mem.Allocator, public_key: [64]u8) ![]const u8 {
    const hash = try primitives.Hash.keccak256(allocator, &public_key);
    defer allocator.free(hash);

    const address_bytes = hash[12..32];
    var hex_addr = try allocator.alloc(u8, 42);
    hex_addr[0] = '0';
    hex_addr[1] = 'x';
    _ = std.fmt.bufPrint(hex_addr[2..], "{s}", .{std.fmt.fmtSliceHexLower(address_bytes)}) catch unreachable;
    return hex_addr;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== Transaction Signing ===\n\n", .{});

    // Generate keypair
    var private_key: [32]u8 = undefined;
    crypto.secp256k1.generatePrivateKey(&private_key);
    const public_key = try crypto.secp256k1.derivePublicKey(private_key);
    const signer_address = try deriveAddress(allocator, public_key);
    defer allocator.free(signer_address);

    std.debug.print("Signer address: {s}\n", .{signer_address});

    // Create unsigned transaction
    const nonce: u64 = 5;
    const gas_price: u64 = 20_000_000_000; // 20 Gwei
    const gas_limit: u64 = 21_000;
    const to = "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb";
    const value: u128 = 1_000_000_000_000_000_000; // 1 ETH
    const chain_id: u64 = 1; // Mainnet

    std.debug.print("\n=== Unsigned Transaction ===\n", .{});
    std.debug.print("  Nonce: {d}\n", .{nonce});
    std.debug.print("  Gas Price: {d} wei (20 Gwei)\n", .{gas_price});
    std.debug.print("  Gas Limit: {d}\n", .{gas_limit});
    std.debug.print("  To: {s}\n", .{to});
    std.debug.print("  Value: {d} wei (1 ETH)\n", .{value});
    std.debug.print("  Data: 0x\n", .{});
    std.debug.print("  Chain ID: {d} (Mainnet)\n", .{chain_id});

    // For simplicity, create a simple message hash representing the transaction
    // In production, you would RLP encode: [nonce, gasPrice, gasLimit, to, value, data, chainId, 0, 0]
    const tx_data = try std.fmt.allocPrint(
        allocator,
        "nonce={d},gasPrice={d},gasLimit={d},to={s},value={d},chainId={d}",
        .{ nonce, gas_price, gas_limit, to, value, chain_id },
    );
    defer allocator.free(tx_data);

    const tx_hash = try primitives.Hash.keccak256(allocator, tx_data);
    defer allocator.free(tx_hash);

    std.debug.print("\nTransaction hash: {s}\n", .{std.fmt.fmtSliceHexLower(tx_hash)});

    // Sign transaction
    const signature = try crypto.secp256k1.sign(tx_hash[0..32].*, private_key);

    std.debug.print("\n=== Signature ===\n", .{});
    std.debug.print("  r: {s}\n", .{std.fmt.fmtSliceHexLower(&signature.r)});
    std.debug.print("  s: {s}\n", .{std.fmt.fmtSliceHexLower(&signature.s)});
    std.debug.print("  v (raw): {d}\n", .{signature.v});

    // Apply EIP-155 v value: v = chainId * 2 + 35 + recovery_id
    const recovery_id: u64 = signature.v - 27;
    const eip155_v = chain_id * 2 + 35 + recovery_id;
    std.debug.print("  v (EIP-155): {d} (with chain ID)\n", .{eip155_v});

    std.debug.print("\n=== Signed Transaction ===\n", .{});
    std.debug.print("Transaction is now ready for broadcast\n", .{});

    // Verify signature by recovering sender
    std.debug.print("\n=== Sender Recovery ===\n", .{});

    const recovered_public_key = try crypto.secp256k1.recoverPublicKey(signature, tx_hash[0..32].*);
    const recovered_address = try deriveAddress(allocator, recovered_public_key);
    defer allocator.free(recovered_address);

    std.debug.print("Recovered address: {s}\n", .{recovered_address});
    std.debug.print("Matches signer: {s}\n", .{if (std.mem.eql(u8, recovered_address, signer_address)) "✓ Yes" else "✗ No"});

    // Demonstrate EIP-155 replay protection
    std.debug.print("\n=== EIP-155 Replay Protection ===\n", .{});

    const mainnet_v = 1 * 2 + 35 + recovery_id; // Chain ID 1
    const sepolia_v = 11155111 * 2 + 35 + recovery_id; // Chain ID 11155111

    std.debug.print("Same signature, different chains:\n", .{});
    std.debug.print("  Mainnet v: {d}\n", .{mainnet_v});
    std.debug.print("  Sepolia v: {d}\n", .{sepolia_v});
    std.debug.print("Different v values prevent replay attacks across chains\n", .{});

    // Demonstrate pre-EIP-155 (legacy)
    std.debug.print("\n=== Pre-EIP-155 (Legacy) ===\n", .{});
    std.debug.print("Legacy v: {d} (no chain ID protection)\n", .{signature.v});
    std.debug.print("Vulnerable to replay attacks across chains\n", .{});
}

File: examples/crypto/secp256k1/transaction-signing.ts
/**
 * Transaction signing and verification
 *
 * Demonstrates:
 * - Legacy transaction signing (pre-EIP-1559)
 * - EIP-155 replay protection
 * - Transaction hash computation
 * - Sender recovery from transaction signature
 * - Transaction verification
 */

import * as Secp256k1 from "../../../src/crypto/Secp256k1/index.js";
import { keccak256 } from "../../../src/primitives/Hash/BrandedHash/keccak256.js";

// Helper: Derive Ethereum address from public key
function deriveAddress(publicKey: Uint8Array): string {
	const hash = keccak256(publicKey);
	const addressBytes = hash.slice(12);
	return "0x" + Buffer.from(addressBytes).toString("hex");
}

// Helper: Simple transaction message (in production, use proper RLP encoding)
function createTransactionMessage(tx: {
	nonce: bigint;
	gasPrice: bigint;
	gasLimit: bigint;
	to: string;
	value: bigint;
	data: Uint8Array;
	chainId: bigint;
}): Uint8Array {
	// For demonstration, create a simple message (production would use RLP)
	const message = `nonce=${tx.nonce},gasPrice=${tx.gasPrice},gasLimit=${tx.gasLimit},to=${tx.to},value=${tx.value},chainId=${tx.chainId}`;
	return new TextEncoder().encode(message);
}

console.log("=== Transaction Signing ===\n");

// Generate keypair
const privateKey = new Uint8Array(32);
crypto.getRandomValues(privateKey);
const publicKey = Secp256k1.derivePublicKey(privateKey);
const signerAddress = deriveAddress(publicKey);

console.log("Signer address:", signerAddress);

// Create unsigned transaction
const unsignedTx = {
	nonce: 5n,
	gasPrice: 20_000_000_000n, // 20 Gwei
	gasLimit: 21_000n, // Standard transfer
	to: "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb",
	value: 1_000_000_000_000_000_000n, // 1 ETH
	data: new Uint8Array(),
	chainId: 1n, // Mainnet
};

console.log("\n=== Unsigned Transaction ===");
console.log("  Nonce:", unsignedTx.nonce);
console.log("  Gas Price:", unsignedTx.gasPrice, "wei (20 Gwei)");
console.log("  Gas Limit:", unsignedTx.gasLimit);
console.log("  To:", unsignedTx.to);
console.log("  Value:", unsignedTx.value, "wei (1 ETH)");
console.log("  Data: 0x");
console.log("  Chain ID:", unsignedTx.chainId, "(Mainnet)");

// Create transaction message
const txMessage = createTransactionMessage(unsignedTx);
console.log(
	"\nTransaction message:",
	Buffer.from(txMessage).toString("hex").slice(0, 64) + "...",
);

// Hash transaction
const txHash = keccak256(txMessage);
console.log("Transaction hash:", Buffer.from(txHash).toString("hex"));

// Sign transaction
const signature = Secp256k1.sign(txHash, privateKey);
console.log("\n=== Signature ===");
console.log("  r:", Buffer.from(signature.r).toString("hex"));
console.log("  s:", Buffer.from(signature.s).toString("hex"));
console.log("  v (raw):", signature.v);

// Apply EIP-155 v value: v = chainId * 2 + 35 + recovery_id
const recoveryId = signature.v - 27;
const eip155V = unsignedTx.chainId * 2n + 35n + BigInt(recoveryId);
console.log("  v (EIP-155):", eip155V, "(with chain ID)");

// Create signed transaction
const signedTx = {
	...unsignedTx,
	r: signature.r,
	s: signature.s,
	v: Number(eip155V),
};

console.log("\n=== Signed Transaction ===");
console.log("Transaction is now ready for broadcast");

// Verify signature by recovering sender
console.log("\n=== Sender Recovery ===");

const recoveredPublicKey = Secp256k1.recoverPublicKey(signature, txHash);
const recoveredAddress = deriveAddress(recoveredPublicKey);

console.log("Recovered address:", recoveredAddress);
console.log(
	"Matches signer:",
	recoveredAddress === signerAddress ? "✓ Yes" : "✗ No",
);

// Demonstrate EIP-155 replay protection
console.log("\n=== EIP-155 Replay Protection ===");

const mainnetV = 1n * 2n + 35n + BigInt(recoveryId); // Chain ID 1
const sepoliaV = 11155111n * 2n + 35n + BigInt(recoveryId); // Chain ID 11155111

console.log("Same signature, different chains:");
console.log("  Mainnet v:", mainnetV);
console.log("  Sepolia v:", sepoliaV);
console.log("Different v values prevent replay attacks across chains");

// Demonstrate pre-EIP-155 (legacy)
console.log("\n=== Pre-EIP-155 (Legacy) ===");
const legacyV = signature.v; // 27 or 28
console.log("Legacy v:", legacyV, "(no chain ID protection)");
console.log("Vulnerable to replay attacks across chains");

// Transaction verification function
function verifyTransaction(
	tx: typeof signedTx,
	expectedSigner: string,
): boolean {
	try {
		// Reconstruct signing message
		const message = createTransactionMessage(tx);
		const hash = keccak256(message);

		// Extract signature (convert EIP-155 v back to recovery ID)
		const chainId = tx.chainId;
		const recoveryId = tx.v - Number(chainId * 2n + 35n);
		const sig = {
			r: tx.r,
			s: tx.s,
			v: recoveryId + 27,
		};

		// Recover signer
		const pubKey = Secp256k1.recoverPublicKey(sig, hash);
		const signer = deriveAddress(pubKey);

		return signer === expectedSigner;
	} catch {
		return false;
	}
}

console.log("\n=== Transaction Verification ===");
const isValid = verifyTransaction(signedTx, signerAddress);
console.log("Transaction valid:", isValid ? "✓ Yes" : "✗ No");

const isInvalid = verifyTransaction(
	signedTx,
	"0x0000000000000000000000000000000000000000",
);
console.log("Wrong signer:", isInvalid ? "✓ Valid" : "✗ Invalid (expected)");
